### PR TITLE
IM array and nested object validation, and element counting

### LINKED ISSUE #873: IM Element Counter Not Counting Correctly
IM element counter is not correctly functioning when json input data is being validated due to the recursive nature of the json validation method.

## Expected Behavior
The IM counts totals of valid elements, invalid elements, elements fixed, and total elements and reports the numbers to OM. The counts are currently not working correctly.

## Current Behavior
The `_validate_json_element` method is recursive to allow for checking nested dictionary elements. The dictionary tracking the values for the different elements to report to OM is instantiated in this method but gets reset when there is a nested variable that is validated.

## Possible Solution
One solution I quickly tested before writing this up was making the element trackers attributes of the IM and then adding a responsibility to the IM's `flush_pool()` method to also reset these values and it seemed to work well.

## Steps to Reproduce
Outlined by @JoeWaddell in PR #828. He was testing the new CLI verbose messaging and saw that an element was fixed during the validation process but the end result of the validation showed a count 0 elements fixed.

### PR DESCRIPTION
**Pull request by ltp23**
In this PR I modified/refactored several existing methods by creating second versions instead of doing so directly. I also added some new methods. I edited some metadata that got flagged. 

The total number of validated elements has increased from about 530K to 1.7mil elements.

I'm working on finalizing the remaining unit tests.

Attempt to close #873 and #683 

---

Update 1: I have modified in-place all the methods that needed to be changed for this PR. Updated all the relevant unit tests and wrote new ones.

---

Update 2: Fixed merged conflicts to make the PR work again. Compiled a list of follow-up tasks in issue #1401.

---

Update 3: Changed the default packing density to 200. Removed the active N percent and denitrification rate input variables.

---

Update 4: Removed issue #1401. Will replace with individual issues for each action item.

### CODE REVIEW

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py]:
```suggestion
    def _validate_json_element(
```
flake8 has no issues with this method, why is there noqa here?

Comment (ltp23) [reply RUFAS/input_manager.py]: Removed now.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py]:
```suggestion
```
This check is redundant. An invalid file type will be caught and handled when Input Manager attempts to load the file.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py:741]:
```suggestion
        elements_counter: ElementsCounter,
```
I think it would be better to move `ElementsCounter` to the top of the file, then type hit it normally.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py]:
```suggestion
```
This method is defined twice.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py]: I think `ElementState` and `ElementCounter` are adding more than there need to be. Here are some changes that I think would make things simpler:
- Remove `decrement()` in `ElementCounter`. It isn't used at all, so it seems like it isn't needed. Also `_check_negative_counts()` could be removed.
- Remove `ElementState` and `increment()`. Instead have there be three methods in `ElementCounter`, one for adding a valid element, one for adding a fixed element, and one for adding an invalid element.

Alternatively, `ElementCounter` could be removed entirely and replaced with a few methods in Input Manager used to increment the counts of valid, fixed, and invalid elements.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py:946]: If the input data is a dict, it should be validated with the method dedicated to validating dict inputs.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py:889]: The type should not be Any, it should be a list.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py]: This method is extremely similar to `_validate_dict_element()`. Please combine it with that method.

**Review by Isgotkowitz — CHANGES_REQUESTED** [review]
Requested changes in the Input Manager. I think there are some things that could be done to make it simpler, though some of the suggestions may not apply after the changes from main are merged in.

File: RUFAS/input_manager.py
Comment (ltp23) [reply RUFAS/input_manager.py]: It's been removed now.

File: RUFAS/input_manager.py
Comment (ltp23) [reply RUFAS/input_manager.py:741]: Yes, this would help with typing inside the InputManager class, but within the ElementsCounter class, I would still need to use quotation marks around the type (like the add method).

File: RUFAS/input_manager.py
Comment (ltp23) [reply RUFAS/input_manager.py]: It's been removed now. I transferred this code to an earlier PR, so I thought it wouldn't duplicate but it did, probably because of the differences in location.

File: RUFAS/input_manager.py
Comment (ltp23) [reply RUFAS/input_manager.py]: I can remove `decrement()` and `_check_negative_counts()`. I would hold off on removing the class because I like some version of separation of concerns here and the decoupling helps with getting the counts from independent elements more easily, like when adding additional variables to the pool.

File: RUFAS/input_manager.py
Comment (ltp23) [reply RUFAS/input_manager.py:889]: The type still needs to be checked inside the method to make sure it is a list.

File: RUFAS/input_manager.py
Comment (ltp23) [reply RUFAS/input_manager.py]: I have tried the following and it works: remove the `dict_type_validator()` method, and use `_validate_input_by_type()` method directly in `_populate_pool()` and `_add_variable_to_pool()`. 

I think this approach makes sense because each element will be funneled through the `_validate_input_by_type()` method first and go to the appropriate type validator.

Changing the code is fairly simple, but fixing the broken tests and/or writing new ones in their place takes some non-trivial amount of work and I can't do that right now. So, I will leave that as one of the follow-up housekeeping tasks.

File: RUFAS/input_manager.py
Comment (ltp23) [reply RUFAS/input_manager.py:946]: I would revisit this point when combining the two `_dict_type_validator()` and `_object_type_validator()`.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [reply RUFAS/input_manager.py]: Ok, that sounds like a plan.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py]:
```suggestion
            The name of the data element just under the level of properties_blob_key.
```
Two things here:
1. It would be incorrect to say that this is came from a JSON, it isn't known in the context of this function where it came from.
2. Could please reword this description or give some examples? I am not sure what you mean by this.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py]:
```suggestion
                    "Validation: array length less than minimum",
```
"array container" makes it sounds like it is something that contains arrays.

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py]:
```suggestion
                    "Validation: array length greater than maximum",
```

File: RUFAS/input_manager.py
Comment (Isgotkowitz) [inline RUFAS/input_manager.py]:
```suggestion
    VALID = "valid"
    INVALID = "invalid"
    FIXED = "fixed"
```
Definitely more redundant, but is clearer.

File: input/metadata/properties/default.json
Comment (Isgotkowitz) [inline input/metadata/properties/default.json]: Where is this number coming from?

File: input/metadata/properties/default.json
Comment (Isgotkowitz) [inline input/metadata/properties/default.json]:
```suggestion
        },
```
These should be removed from inputs, instead of being allowed in.

File: input/metadata/properties/default.json
Comment (Isgotkowitz) [inline input/metadata/properties/default.json]:
```suggestion
          "pattern": "^(freestall|tiestall|open lot|compost bedded pack barn)$",
```

**Review by Isgotkowitz — APPROVED** [review]

File: input/metadata/properties/default.json
Comment (JoeWaddell) [reply input/metadata/properties/default.json]: Looks like from the conflict between the default and the former minimum, but confirmation that one or the other is correct from a SME would be best. I couldn't find it in the Feed design doc, and checking some sources online, 14 kg / m^3 seems low.

File: RUFAS/input_manager.py
Comment (JoeWaddell) [reply RUFAS/input_manager.py]: Strings make more intuitive sense to me as well, but for my own edification, is there any benefit in using the existing ints?

**Review by JoeWaddell — COMMENTED** [review]
Assuming @Isgotkowitz 's remaining comments are addressed prior to merge, this looks good. 

Given the size of this PR I agree with pushing some of the cleanup to Issues #1335 and #1401

**Comment by ltp23** [comment]
> Looks like from the conflict between the default and the former minimum, but confirmation that one or the other is correct from a SME would be best. I couldn't find it in the Feed design doc, and checking some sources online, 14 kg / m^3 seems low.

There seems to be a mismatch between the metadata and the input data also. I saw that all the feed json files use 14 so I thought that I would update the metadata. But of course, this can be changed with further confirmation.

---

Update: I changed the default to 200.

File: input/metadata/properties/default.json
Comment (ltp23) [reply input/metadata/properties/default.json]: Ok I will remove these in the metadata and input json files.

**Review by JoeWaddell — APPROVED** [review]
LGTM

**Pull request merged**
